File: py_src/taskito/proxies/metrics.py
"""Proxy reconstruction performance metrics."""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Any


@dataclass
class HandlerMetrics:
    """Per-handler reconstruction metrics."""

    name: str
    total_reconstructions: int = 0
    total_errors: int = 0
    total_cleanup_errors: int = 0
    total_checksum_failures: int = 0
    total_duration_ms: float = 0.0
    max_duration_ms: float = 0.0
    _durations: list[float] = field(default_factory=list)

    @property
    def avg_duration_ms(self) -> float:
        if not self._durations:
            return 0.0
        return sum(self._durations) / len(self._durations)

    @property
    def p95_duration_ms(self) -> float:
        if not self._durations:
            return 0.0
        sorted_d = sorted(self._durations)
        idx = int(len(sorted_d) * 0.95)
        return sorted_d[min(idx, len(sorted_d) - 1)]


class ProxyMetrics:
    """Thread-safe accumulator for proxy reconstruction metrics."""

    def __init__(self) -> None:
        self._handlers: dict[str, HandlerMetrics] = {}
        self._lock = threading.Lock()

    def _get_handler(self, name: str) -> HandlerMetrics:
        if name not in self._handlers:
            self._handlers[name] = HandlerMetrics(name=name)
        return self._handlers[name]

    def record_reconstruction(self, handler_name: str, duration_ms: float) -> None:
        with self._lock:
            h = self._get_handler(handler_name)
            h.total_reconstructions += 1
            h.total_duration_ms += duration_ms
            h.max_duration_ms = max(h.max_duration_ms, duration_ms)
            h._durations.append(duration_ms)

    def record_error(self, handler_name: str) -> None:
        with self._lock:
            self._get_handler(handler_name).total_errors += 1

    def record_cleanup_error(self, handler_name: str) -> None:
        with self._lock:
            self._get_handler(handler_name).total_cleanup_errors += 1

    def record_checksum_failure(self, handler_name: str) -> None:
        with self._lock:
            self._get_handler(handler_name).total_checksum_failures += 1

    def to_list(self) -> list[dict[str, Any]]:
        """Return metrics as a list of dicts."""
        with self._lock:
            return [
                {
                    "handler": h.name,
                    "total_reconstructions": h.total_reconstructions,
                    "total_errors": h.total_errors,
                    "total_cleanup_errors": h.total_cleanup_errors,
                    "total_checksum_failures": h.total_checksum_failures,
                    "total_duration_ms": round(h.total_duration_ms, 2),
                    "avg_duration_ms": round(h.avg_duration_ms, 2),
                    "max_duration_ms": round(h.max_duration_ms, 2),
                    "p95_duration_ms": round(h.p95_duration_ms, 2),
                }
                for h in self._handlers.values()
            ]

File: sdks/node/src/proxies/metrics.ts
// Process-wide accumulator for proxy reconstruction metrics, surfaced by the
// dashboard's /api/proxy-stats endpoint (snake_case contract shape).

interface HandlerMetrics {
  reconstructions: number;
  errors: number;
  cleanupErrors: number;
  checksumFailures: number;
  totalDurationMs: number;
  maxDurationMs: number;
  durations: number[];
}

/** One handler's stats in the dashboard wire shape. */
export interface ProxyHandlerStats {
  handler: string;
  total_reconstructions: number;
  total_errors: number;
  total_cleanup_errors: number;
  total_checksum_failures: number;
  total_duration_ms: number;
  avg_duration_ms: number;
  max_duration_ms: number;
  p95_duration_ms: number;
}

/** Cap the retained duration samples so long-lived processes stay bounded. */
const MAX_SAMPLES = 1000;

export class ProxyMetrics {
  private readonly handlers = new Map<string, HandlerMetrics>();

  private handler(name: string): HandlerMetrics {
    let entry = this.handlers.get(name);
    if (!entry) {
      entry = {
        reconstructions: 0,
        errors: 0,
        cleanupErrors: 0,
        checksumFailures: 0,
        totalDurationMs: 0,
        maxDurationMs: 0,
        durations: [],
      };
      this.handlers.set(name, entry);
    }
    return entry;
  }

  recordReconstruction(handlerName: string, durationMs: number): void {
    const entry = this.handler(handlerName);
    entry.reconstructions += 1;
    entry.totalDurationMs += durationMs;
    entry.maxDurationMs = Math.max(entry.maxDurationMs, durationMs);
    entry.durations.push(durationMs);
    if (entry.durations.length > MAX_SAMPLES) {
      entry.durations.shift();
    }
  }

  recordError(handlerName: string): void {
    this.handler(handlerName).errors += 1;
  }

  recordCleanupError(handlerName: string): void {
    this.handler(handlerName).cleanupErrors += 1;
  }

  recordChecksumFailure(handlerName: string): void {
    this.handler(handlerName).checksumFailures += 1;
  }

  toList(): ProxyHandlerStats[] {
    return [...this.handlers.entries()].map(([name, h]) => {
      const sorted = [...h.durations].sort((a, b) => a - b);
      const p95 = sorted.length > 0 ? sorted[Math.floor(sorted.length * 0.95)] : 0;
      return {
        handler: name,
        total_reconstructions: h.reconstructions,
        total_errors: h.errors,
        total_cleanup_errors: h.cleanupErrors,
        total_checksum_failures: h.checksumFailures,
        total_duration_ms: round2(h.totalDurationMs),
        avg_duration_ms: round2(h.reconstructions > 0 ? h.totalDurationMs / h.reconstructions : 0),
        max_duration_ms: round2(h.maxDurationMs),
        p95_duration_ms: round2(p95 ?? sorted[sorted.length - 1] ?? 0),
      };
    });
  }

  reset(): void {
    this.handlers.clear();
  }
}

const round2 = (value: number): number => Math.round(value * 100) / 100;

/** Shared per-process accumulator every `Proxies` registry records into. */
export const proxyMetrics = new ProxyMetrics();
